File: client-windows/internal/updater/updater.go
// Package updater downloads and applies over-the-air updates of the combined
// rootaika client binary. The pure parts (NeedsUpdate, Download) are testable on
// any platform; the self-swap helper is split into apply_windows.go and
// apply_other.go.
package updater

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"rootaika/client-windows/internal/version"
)

// Plan is the server-declared update target: which release tag to fetch, which
// asset within it, and the expected SHA256 of that asset. An empty Version means
// no update is desired.
type Plan struct {
	Version  string
	Artifact string
	SHA256   string
}

// retry policy for the download. Matches the api client stance: retry network
// and 5xx, never 4xx, and a SHA256 mismatch is terminal (handled in Download).
const (
	maxAttempts = 4
	baseBackoff = 500 * time.Millisecond
	maxBackoff  = 5 * time.Second
)

// NeedsUpdate reports whether the plan asks for a version different from the one
// currently running. Comparison is plain string inequality (no semver ordering),
// so any difference, including a downgrade, triggers an update. An empty desired
// version or an incomplete plan never triggers.
func NeedsUpdate(current string, p Plan) bool {
	if p.Version == "" || p.Artifact == "" || p.SHA256 == "" {
		return false
	}
	return p.Version != current
}

// gitHubBase is the public release-download host. It is a var only so tests can
// point Download at an httptest server; production never overrides it, and the
// owner/repo path segments remain compile-time constants.
const gitHubBase = "https://github.com"

// Download streams the plan's release asset to destPath and verifies its SHA256.
// On any error, including a hash mismatch, destPath is removed and no usable file
// is left behind. The URL is built only from the compile-time-fixed owner/repo
// plus the plan, so the server can never redirect the download elsewhere.
func Download(ctx context.Context, p Plan, destPath string) error {
	return downloadFrom(ctx, gitHubBase, p, destPath)
}

// downloadFrom is Download with an injectable base host for testing. The path is
// always {owner}/{repo}/releases/download/{tag}/{asset}, matching
// version.DownloadURL, so the production call and the test exercise the same
// path construction.
func downloadFrom(ctx context.Context, base string, p Plan, destPath string) error {
	if p.Version == "" || p.Artifact == "" || p.SHA256 == "" {
		return fmt.Errorf("incomplete update plan: %+v", p)
	}
	url := base + "/" + version.GitHubOwner + "/" + version.GitHubRepo +
		"/releases/download/" + p.Version + "/" + p.Artifact

	body, err := fetch(ctx, url)
	if err != nil {
		return err
	}
	defer body.Close()

	if err := streamToFile(body, destPath, p.SHA256); err != nil {
		_ = os.Remove(destPath)
		return err
	}
	return nil
}

// fetch performs the GET with retry/backoff, returning the response body on a
// 2xx. The caller closes the body. Network errors and 5xx/429 are retried; 4xx
// is terminal.
func fetch(ctx context.Context, url string) (io.ReadCloser, error) {
	client := &http.Client{Timeout: 5 * time.Minute}
	var lastErr error
	for attempt := 0; attempt < maxAttempts; attempt++ {
		if attempt > 0 {
			if err := sleep(ctx, backoff(attempt)); err != nil {
				return nil, err
			}
		}
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
		if err != nil {
			return nil, err
		}
		resp, err := client.Do(req)
		if err != nil {
			if ctx.Err() != nil {
				return nil, err
			}
			lastErr = err
			continue
		}
		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			return resp.Body, nil
		}
		resp.Body.Close()
		retryable := resp.StatusCode >= 500 || resp.StatusCode == http.StatusTooManyRequests
		lastErr = fmt.Errorf("download %s failed with %s", url, resp.Status)
		if !retryable {
			return nil, lastErr
		}
	}
	return nil, fmt.Errorf("download failed after %d attempts: %w", maxAttempts, lastErr)
}

// streamToFile copies src to destPath while hashing, then fails if the hash does
// not match wantHex (case-insensitive). The file is synced before returning so
// the subsequent swap sees the full contents.
func streamToFile(src io.Reader, destPath, wantHex string) error {
	out, err := os.OpenFile(destPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o755)
	if err != nil {
		return err
	}
	hasher := sha256.New()
	if _, err := io.Copy(out, io.TeeReader(src, hasher)); err != nil {
		out.Close()
		return err
	}
	if err := out.Sync(); err != nil {
		out.Close()
		return err
	}
	if err := out.Close(); err != nil {
		return err
	}
	got := hex.EncodeToString(hasher.Sum(nil))
	if !strings.EqualFold(got, strings.TrimSpace(wantHex)) {
		return fmt.Errorf("sha256 mismatch: got %s want %s", got, wantHex)
	}
	return nil
}

func backoff(attempt int) time.Duration {
	delay := baseBackoff << (attempt - 1)
	if delay > maxBackoff {
		delay = maxBackoff
	}
	return delay
}

func sleep(ctx context.Context, d time.Duration) error {
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

File: client-macos/Sources/RootaikaMac/Updater.swift
import Foundation
import CryptoKit

/// OTA self-update, mirroring the Windows updater: the server's config names a
/// desired release (version tag / asset name / SHA256); the daemon downloads
/// the GitHub release asset, verifies the hash, renames it over the installed
/// binary (rename(2) is atomic; running processes keep the old inode), and
/// exits so launchd's KeepAlive restarts daemon and agent on the new version.
enum Updater {
    struct Plan: Equatable {
        let version: String
        let artifact: String
        let sha256: String

        var isComplete: Bool {
            !version.isEmpty && !artifact.isEmpty && !sha256.isEmpty
        }
    }

    enum UpdateError: Error, CustomStringConvertible {
        case incompletePlan
        case badURL
        case status(Int)
        case sha256Mismatch(got: String, want: String)
        case renameFailed(errno: Int32)

        var description: String {
            switch self {
            case .incompletePlan: return "incomplete update plan"
            case .badURL: return "cannot build download URL"
            case .status(let code): return "download failed with status \(code)"
            case .sha256Mismatch(let got, let want): return "sha256 mismatch: got \(got) want \(want)"
            case .renameFailed(let errno): return "rename failed: errno \(errno)"
            }
        }
    }

    /// Same policy as the Windows updater and the API client: retry network
    /// errors, 5xx and 429; 4xx and a hash mismatch are terminal.
    private static let maxAttempts = 4
    private static let baseBackoff: TimeInterval = 0.5
    private static let maxBackoff: TimeInterval = 5.0

    /// Plain string inequality (no semver ordering): any difference, including
    /// a downgrade, triggers an update. An incomplete plan never triggers.
    static func needsUpdate(current: String, plan: Plan) -> Bool {
        guard plan.isComplete else { return false }
        return plan.version != current
    }

    /// {base}/{owner}/{repo}/releases/download/{tag}/{asset}, owner/repo fixed
    /// at compile time.
    static func downloadURL(plan: Plan, base: String = "https://github.com") -> URL? {
        URL(string: "\(base)/\(Version.gitHubOwner)/\(Version.gitHubRepo)/releases/download/\(plan.version)/\(plan.artifact)")
    }

    static func sha256Hex(of data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    /// Download the plan's release asset to `destination` and verify its
    /// SHA256. On any failure no usable file is left behind.
    static func download(plan: Plan, to destination: URL, base: String = "https://github.com") async throws {
        guard plan.isComplete else { throw UpdateError.incompletePlan }
        guard let url = downloadURL(plan: plan, base: base) else { throw UpdateError.badURL }

        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 300
        config.timeoutIntervalForResource = 300
        let session = URLSession(configuration: config)

        var lastError: Error = UpdateError.status(0)
        for attempt in 1...maxAttempts {
            if attempt > 1 {
                let delay = min(maxBackoff, baseBackoff * pow(2.0, Double(attempt - 2)))
                try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            }
            do {
                let (data, response) = try await session.data(from: url)
                guard let http = response as? HTTPURLResponse else {
                    lastError = UpdateError.status(0)
                    continue
                }
                if http.statusCode == 429 || (500...599).contains(http.statusCode) {
                    lastError = UpdateError.status(http.statusCode)
                    continue
                }
                guard (200...299).contains(http.statusCode) else {
                    throw UpdateError.status(http.statusCode) // terminal 4xx
                }
                let got = sha256Hex(of: data)
                let want = plan.sha256.trimmingCharacters(in: .whitespaces).lowercased()
                guard got == want else {
                    throw UpdateError.sha256Mismatch(got: got, want: want) // terminal
                }
                try data.write(to: destination, options: .atomic)
                try FileManager.default.setAttributes([.posixPermissions: 0o755], ofItemAtPath: destination.path)
                return
            } catch let error as UpdateError {
                try? FileManager.default.removeItem(at: destination)
                throw error
            } catch is CancellationError {
                throw lastError
            } catch {
                lastError = error // transport error: retry
            }
        }
        try? FileManager.default.removeItem(at: destination)
        throw lastError
    }

    /// Atomically replace the live binary with the staged download.
    static func apply(staged: URL, target: URL) throws {
        let result = rename(staged.path, target.path)
        guard result == 0 else {
            let code = errno
            try? FileManager.default.removeItem(at: staged)
            throw UpdateError.renameFailed(errno: code)
        }
    }
}
